aws/rust-runtime/aws-inlineable/src/glacier_interceptors.rs

```diff
@@ -0,0 +1,454 @@
+        request.headers_mut().insert(
+            HeaderName::from_static(X_AMZ_CONTENT_SHA256),
+            complete_hash.parse().expect("hash must be valid header"),
+        );
+    }
+    Ok(())
+}
+
+const MEGABYTE: usize = 1024 * 1024;
+async fn compute_hashes(
+    body: SdkBody,
+    chunk_size: usize,
+) -> Result<(Digest, Vec<Digest>), byte_stream::error::Error> {
+    let mut hashes = vec![];
+    let mut remaining_in_chunk = chunk_size;
+    let mut body = ByteStream::new(body);
+    let mut local = Context::new(&SHA256);
+    let mut full_body = Context::new(&SHA256);
+    let mut segmented = SegmentedBuf::new();
+    while let Some(data) = body.try_next().await? {
+        segmented.push(data);
+        while segmented.has_remaining() {
+            let next = segmented.chunk();
+            let len = next.len().min(remaining_in_chunk);
+            local.update(&next[..len]);
+            full_body.update(&next[..len]);
+            segmented.advance(len);
+            remaining_in_chunk -= len;
+            if remaining_in_chunk == 0 {
+                hashes.push(local.finish());
+                local = Context::new(&SHA256);
+                remaining_in_chunk = chunk_size;
+            }
+        }
+    }
+    if remaining_in_chunk != chunk_size || hashes.is_empty() {
+        hashes.push(local.finish());
+    }
+    Ok((full_body.finish(), hashes))
+}
+
+/// Compute the glacier tree hash for a vector of hashes.
+///
+/// Adjacent hashes are combined into a single hash. This process occurs recursively until only 1 hash remains.
+///
+/// See <https://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html> for more information.
+fn compute_hash_tree(mut hashes: Vec<Digest>) -> Digest {
+    assert!(
+        !hashes.is_empty(),
+        "even an empty file will produce a digest. this function assumes that hashes is non-empty"
+    );
+    while hashes.len() > 1 {
+        let next = hashes.chunks(2).into_iter().map(|chunk| match *chunk {
+            [left, right] => {
+                let mut ctx = Context::new(&SHA256);
+                ctx.update(left.as_ref());
+                ctx.update(right.as_ref());
+                ctx.finish()
+            }
+            [last] => last,
+            _ => unreachable!(),
+        });
+        hashes = next.collect();
+    }
+    hashes[0]
+}
+
+#[cfg(test)]
+mod account_id_autofill_tests {
+    use super::*;
+    use aws_smithy_runtime_api::type_erasure::TypedBox;
+
+    #[test]
+    fn autofill_account_id() {
+        #[derive(Debug)]
+        struct SomeInput {
+            account_id: Option<String>,
+        }
+        impl GlacierAccountId for SomeInput {
+            fn account_id_mut(&mut self) -> &mut Option<String> {
+                &mut self.account_id
+            }
+        }
+
+        let mut cfg = ConfigBag::base();
+        let mut context =
+            InterceptorContext::<()>::new(TypedBox::new(SomeInput { account_id: None }).erase());
```

Comment:
We create structs for this so much in testing it'd be cool to have a derive macro for `From<T> for TypedBox`. I'm guessing those require us to use `syn` tho.

Reply:
Implementing a generic `From<T> for Whatever` seems to always conflict with the `From<T> for T` that's built into the language, so that's not really feasible.

Reply:
you don't need a derive macro, just `erase!(...)` which is shorthand for

```
macro_rules! erase {
 ($input: expr) => { TypedBox::new($input).erase() }
```

Reply:
I like that macro idea.